From 40dc4aae8b9689d36412485017f1cb18a140c9bc Mon Sep 17 00:00:00 2001
From: Yongxing Mou <yongxing.mou@oss.qualcomm.com>
Date: Mon, 25 Aug 2025 22:16:24 +0800
Subject: [PATCH 2/6] FROMLIST: drm/msm/dp: Add MST stream support for SA8775P
 DP controller 0 and 1

This change enables SA8775P support for Multi-Stream Transport (MST),
allowing each controller to handle up to two or four DisplayPort streams.
As all necessary code for MST support was already implemented in the
previous series of patches.

Signed-off-by: Yongxing Mou <yongxing.mou@oss.qualcomm.com>
Upstream-Status: Submitted [https://lore.kernel.org/all/20250825-msm-dp-mst-v3-38-01faacfcdedd@oss.qualcomm.com/]
---
 drivers/gpu/drm/msm/dp/dp_display.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/msm/dp/dp_display.c b/drivers/gpu/drm/msm/dp/dp_display.c
index 5ac957010843..6bd89a07445c 100644
--- a/drivers/gpu/drm/msm/dp/dp_display.c
+++ b/drivers/gpu/drm/msm/dp/dp_display.c
@@ -108,6 +108,11 @@ enum dp_mst_intf {
 	INTF_MAX
 };
 
+static const unsigned int stream_intf_map_qcs8300[][DP_STREAM_MAX] = {
+	{INTF_0, INTF_3, INTF_6, INTF_7},
+	{}
+};
+
 static const unsigned int stream_intf_map_sa_8775p[][DP_STREAM_MAX] = {
 	{INTF_0, INTF_3},
 	{INTF_4, INTF_8},
@@ -115,8 +120,8 @@ static const unsigned int stream_intf_map_sa_8775p[][DP_STREAM_MAX] = {
 };
 
 static const struct msm_dp_desc msm_dp_desc_qcs8300[] = {
-	{ .io_start = 0x0af54000, .id = MSM_DP_CONTROLLER_0, .wide_bus_supported = true, .max_streams = 2,
-	  .intf_map = stream_intf_map_sa_8775p[MSM_DP_CONTROLLER_0],
+	{ .io_start = 0x0af54000, .id = MSM_DP_CONTROLLER_0, .wide_bus_supported = true, .max_streams = 4,
+	  .intf_map = stream_intf_map_qcs8300[MSM_DP_CONTROLLER_0],
 	},
 	{}
 };
@@ -1809,8 +1814,8 @@ void msm_dp_bridge_hpd_notify(struct drm_bridge *bridge,
 	    status == connector_status_connected)
 		msm_dp_irq_hpd_handle(dp, 0);
 	else if (hpd_link_status == ISR_HPD_REPLUG_COUNT) {
-		msm_dp_hpd_plug_handle(dp, 0);
 		msm_dp_hpd_unplug_handle(dp, 0);
+		msm_dp_hpd_plug_handle(dp, 0);
 	} else if ((!msm_dp_display->internal_hpd ||
 		    hpd_link_status == ISR_DISCONNECTED) &&
 		 status == connector_status_disconnected)
-- 
2.34.1

